tests/epyccel/modules/arrays.py

```diff
@@ -1753,3 +1753,304 @@ def multiple_np_linspace():
     y = np.linspace(0, 4, 128)
     z = np.linspace(0, 8, 128)
     return x[0] + y[1] + z[2] + linspace_index
+
+#==============================================================================
+# NUMPY ARRAY DATA TYPE CONVERSION
+#==============================================================================
+
+@template('T1', ['bool[:,:]', 'int[:,:]', 'int8[:,:]', 'int16[:,:]', 'int32[:,:]', 'int64[:,:]', 'float[:,:]', 'float32[:,:]', 'float64[:,:]', 'complex64[:,:]', 'complex128[:,:]'])
+def dtype_convert_to_bool(arr : 'T1'):
+    from numpy import shape, array
+    c = array(arr, dtype='bool')
+    s = shape(c)
+    return len(s), c[0,0], c[0,1], c[1,0], c[1,1]
+
+@template('T1', ['bool[:,:]', 'int[:,:]', 'int8[:,:]', 'int16[:,:]', 'int32[:,:]', 'int64[:,:]', 'float[:,:]', 'float32[:,:]', 'float64[:,:]', 'complex64[:,:]', 'complex128[:,:]'])
+def dtype_convert_to_int8(arr : 'T1'):
+    from numpy import shape, array
+    c = array(arr, dtype='int8')
+    s = shape(c)
+    return len(s), c[0,0], c[0,1], c[1,0], c[1,1]
```

Comment:
@EmilyBourne There is a lot of code duplication in this file. Clearly the problem is that Pyccel must be able to infer the type of the `c` array (and `d` array further down). Does this suggest that we need templated variables in addition to templated types?

Reply:
I'm not sure what you mean by "templated variables". The problem here is mostly on the Python side. I think Pyccel would happily translate:
```python
@template('T1', ['bool[:,:]', 'int[:,:]', 'int8[:,:]', 'int16[:,:]', 'int32[:,:]', 'int64[:,:]', 'float[:,:]', 'float32[:,:]', 'float64[:,:]', 'complex64[:,:]', 'complex128[:,:]'])
@template('T2', ['bool', 'int', 'int8', 'int16', 'int32', 'int64', 'float', 'float32', 'float64', 'complex64', 'complex128'])
def dtype_convert_to_int8(arr : 'T1'):
    from numpy import shape, array
    c = array(arr, dtype='T2')
    s = shape(c)
    return len(s), c[0,0], c[0,1], c[1,0], c[1,1]
```
But the Python code won't run. And we use the Python code for comparison. Another option is to add an extra parameter with the target type. E.g.:
```python
@template('T1', ['bool[:,:]', 'int[:,:]', 'int8[:,:]', 'int16[:,:]', 'int32[:,:]', 'int64[:,:]', 'float[:,:]', 'float32[:,:]', 'float64[:,:]', 'complex64[:,:]', 'complex128[:,:]'])
@template('T2', ['bool', 'int', 'int8', 'int16', 'int32', 'int64', 'float', 'float32', 'float64', 'complex64', 'complex128'])
def dtype_convert_to_int8(arr : 'T1[:,:]', type_var : 'T2'):
    from numpy import shape, array
    c = array(arr, dtype=type(type_var))
    s = shape(c)
    return len(s), c[0,0], c[0,1], c[1,0], c[1,1]
```

Reply:
The Python interpreter should be happy with
```python
@template('T1', ['bool[:,:]', 'int[:,:]', 'int8[:,:]', 'int16[:,:]', 'int32[:,:]', 'int64[:,:]', 'float[:,:]', 'float32[:,:]', 'float64[:,:]', 'complex64[:,:]', 'complex128[:,:]'])
def dtype_convert_to_other_dtype(arr : 'T1', dtype : type):
    from numpy import shape, array
    c = array(arr, dtype=dtype)
    s = shape(c)
    return len(s), c[0,0], c[0,1], c[1,0], c[1,1]
```
Unfortunately we cannot handle this code in Pyccel. As a workaround we could say that `dtype` is a "templated variable" which can only assume one of the values that we provide through a template. Maybe a proper MyPy annotation exists which provides the necessary information?

Reply:
@VaradRengaraj Feel free to ignore this discussion. This is for us for future developments. It does not affect your PR

Reply:
Hi @EmilyBourne  i know this is not the right place to ask , but can you suggest me any bugs/ backlogs in Pyccel which is not time bound, so that I can work on my own pace. Can you suggest me any such issues?

Reply:
> can you suggest me any bugs/ backlogs in Pyccel which is not time bound

I have replied on discord

Reply:
@EmilyBourne thank you